// videogular2/src/controls/vg-play-pause/vg-play-pause.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var vg_play_pause_1 = require("./vg-play-pause");
var vg_api_1 = require("../../core/services/vg-api");
var vg_states_1 = require("../../core/states/vg-states");
describe('Play/Pause Button', function () {
    var playPause;
    var ref;
    var api;
    beforeEach(function () {
        ref = {
            nativeElement: {
                getAttribute: function (name) {
                    return name;
                }
            }
        };
        api = new vg_api_1.VgAPI();
        api.medias = {
            main: {
                state: vg_states_1.VgStates.VG_PLAYING
            },
            secondary: {
                state: vg_states_1.VgStates.VG_PAUSED
            }
        };
        playPause = new vg_play_pause_1.VgPlayPause(ref, api);
    });
    it('Should get media by id on init', function () {
        spyOn(api, 'getMediaById').and.callFake(function () {
            return {
                volume: 1
            };
        });
        playPause.vgFor = 'test';
        playPause.onPlayerReady();
        expect(api.getMediaById).toHaveBeenCalledWith('test');
    });
    it('Should get state for one media file', function () {
        api.medias = {
            main: {
                state: vg_states_1.VgStates.VG_PLAYING
            }
        };
        playPause.target = api;
        var state = playPause.getState();
        expect(state).toBe(vg_states_1.VgStates.VG_PLAYING);
    });
    describe('onClick (single and multiple media)', function () {
        it('should pause if current state is different play', function () {
            spyOn(api, 'pause').and.callFake(function () { });
            api.medias = {
                main: {
                    state: vg_states_1.VgStates.VG_PLAYING
                }
            };
            playPause.target = api;
            playPause.onClick();
            expect(api.pause).toHaveBeenCalled();
        });
        it('should play if current state is pause', function () {
            spyOn(api, 'play').and.callFake(function () { });
            api.medias = {
                main: {
                    state: vg_states_1.VgStates.VG_PAUSED
                }
            };
            playPause.target = api;
            playPause.onClick();
            expect(api.play).toHaveBeenCalled();
        });
        it('should play if current state is ended', function () {
            spyOn(api, 'play').and.callFake(function () { });
            api.medias = {
                main: {
                    state: vg_states_1.VgStates.VG_ENDED
                }
            };
            playPause.target = api;
            playPause.onClick();
            expect(api.play).toHaveBeenCalled();
        });
    });
});
